### PR TITLE
RFC - Enable Behaviors to participate in marshalling

### PR DESCRIPTION
**Pull request by markstory**
This is a rough sketch of what it could look like for behaviors to be able to participate in marshalling by providing a map of the fields they can marshal. Each behavior could implement an interface opt-ing it into the marshalling process. The behavior would then return an updated map of fields -> marshalling functions.

Each marshalling function is expected to convert the request value into a hydrated object. The unified interface lets database types, associations, and behaviors work the same way.
### TODO
- [ ] Complete tests.
- [ ] Add interface mentioned in comments.
- [ ] Complete documentation.
- [ ] Fix merge conflicts.

These changes are based on those started by @albertcansado in #9130. Given how close 3.3.0 is, and how large of a change this is, I would recommend this go into 3.4.0

### CODE REVIEW

File: src/ORM/Marshaller.php [hunk 1]
```diff
@@ -215,7 +237,7 @@ protected function _validate($data, $options, $isNew)
      */
     protected function _prepareDataAndOptions($data, $options)
     {
-        $options += ['validate' => true];
+        $options += ['validate' => true, 'translations' => true];
```
Comment (lorenzo) [inline src/ORM/Marshaller.php]: Why is this here?

File: src/ORM/Marshaller.php [hunk 2]
```diff
@@ -55,28 +55,59 @@ public function __construct(Table $table)
...
+                $nested['forceNew'] = $options['forceNew'];
+            }
+            $new = function ($value, $entity) use ($assoc, $nested) {
+                $options = $nested + ['associated' => []];
+                return $this->_marshalAssociation($assoc, $value, $options);
+            };
+            $merge = function ($value, $entity) use ($assoc, $nested) {
+                $options = $nested + ['associated' => []];
+                return $this->_mergeAssociation($entity->get($assoc->property()), $assoc, $value, $options);
+            };
```
Comment (lorenzo) [inline src/ORM/Marshaller.php]: I wonder if we can extract these to closures to another method or class. It seems a bit expensive to be creating them in a loop for each entity we want to marshal.

That also opens the questions if we should stop creating a new marshaller instance every time .

File: src/ORM/Marshaller.php (hunk 1 above)
Comment (markstory) [inline src/ORM/Marshaller.php]: I'll remove it.

File: src/ORM/Marshaller.php (hunk 2 above)
Comment (markstory) [inline src/ORM/Marshaller.php]: I had hoped to avoid the extra layer of closures, but the different code paths for new/merge required it. I also had to create closures for each property being marshalled, as I didn't want to end up passing extra parameters to Type instance's marshal methods. It is possible that passing the extra argument is less harmful than the additional function calls though.

Are you suggesting we inline `_mergeAssociation` and `_marshalAssociation` here?

**Comment by markstory** [comment]
Closing as I targeted the wrong branch.
